include/net/socket_can.h

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _SOCKET_CAN_H_
+#define _SOCKET_CAN_H_
+
+#include <device.h>
+#include <net/net_ip.h>
+#include <can.h>
+#include <zephyr/types.h>
+
+/** Max length of socket packet */
+#define CAN_NET_MTU (72)
+
+/**  Protocol family of socket can connection */
+#define CAN_CONTEXT_FAMILY BIT(8)
+
+/**  Context type of socket can connection */
+#define CAN_CONTEXT_TYPE BIT(9)
+
+/**  Context proto for socket can connection */
+#define CAN_CONTEXT_PROTO BIT(10)
+
+/**
+ * @brief socket addr structure for AF_CAN protocol
+ *
+ * Socket address structure for AF_CAN protocol family
+ *
+ */
+struct sockaddr_can {
+	/** protocol family for socket can */
+	sa_family_t can_family;
+	/** net_if interface associated with socket */
+	u8_t can_ifindex;
+};
+
+/**
+ * @brief ioctl call for configuring misc params in can IP.
+ *
+ * This routine provides support for configuring misc params.
+ * *
+ * @param dev       Pointer to the device structure for the driver instance.
+ * @param cmd       type of command.
+ * @param data      data related to cmd to configure parameters.
+ *
+ * @retval 0 if success
+ */
+typedef int (*socket_can_ioctl_t)(struct device *dev, u32_t cmd, void *data);
+
+/**
+ * @brief check matched CAN filter for incoming CAN message.
+ *
+ * This routine matches id with configured filters id and return true or false.
+ * *
+ * @param dev       Pointer to the device structure for the driver instance.
+ * @param pkt       incoming net_pkt
+ *
+ * @retval non zero value if filter matched otherwise returns 0.
+ */
+typedef int (*socket_can_check_filter_matched_t)(struct device *dev,
+						 struct net_pkt *pkt);
+
+/**
+ * @brief socket can driver api structure
+ * Socket CAN driver apis for iface interface functions and ioctl call.
+ *
+ */
+struct socket_can_driver_api_t {
+	/** iface interface for socket apis to send and init function*/
+	struct net_if_api iface_api;
+	/** ioctl call for socket can driver (this will be replaced by */
+	/* setsocketopt call */
+	socket_can_ioctl_t ioctl;
+	/** check matching filter for loopback messages */
+	socket_can_check_filter_matched_t check_matched_filter;
+};
+
+/**
+ * @brief socket can context structure
+ *
+ * socket can driver specific structure.
+ *
+ */
+struct socket_can_context {
+	/** id of CAN instance */
+	u32_t id;
+	/** iface interface reference structure*/
+	struct net_if *iface;
+	/** can device structure reference */
+	struct device *can_dev;
+};
+
+__syscall int socket_can_ioctl(struct device *dev, u32_t cmd, void *data);
```

Comment:
Sockets is purely a POSIX construct. For as long as POSIX doesn't have function socket_can_ioctl(), it doesn't make sense to have it here too.

Reply:
These calls will be replaced by setSocketOpt and getSocketOpt call, currently these calls are not implemented in Zephyr hence need to use customized ioctls calls.

Reply:
getsockopt and setsockopt are there already, afaik.

Reply:
currently these calls returning -1 and have no logic implemented.
zephyr/subsys/net/lib/sockets/sockets.c : 841--
int zsock_getsockopt(int sock, int level, int optname,
             void *optval, socklen_t *optlen)
{
    errno = ENOPROTOOPT;
    return -1;
}

int zsock_setsockopt(int sock, int level, int optname,
             const void *optval, socklen_t optlen)
{
    errno = ENOPROTOOPT;
    return -1;
}